### PR TITLE
feat: add First Value reducer to FT.Aggregate groupby stage

### PR DESCRIPTION
**Pull request by rileydes-improving**
## Summary
Adds the `FIRST_VALUE `reducer function to `FT.AGGREGATE` `GROUPBY` stages, compatible with RediSearch. Supports a simple (non-deterministic) mode and a sorted mode via an optional `BY <field> [ASC|DESC]` clause that returns the value from the record with the minimum or maximum comparison field within each group.

## Changes

- Added `FirstValue` reducer class with simple and sorted (ASC/DESC) modes in `ft_aggregate_exec.cc`.
- Extended `ReducerInfo` in `ft_aggregate_parser.h` with an optional `ArgParser*` for reducers needing custom token parsing, parser dispatches to it when set.
- Added `ParseFirstValueArgs` custom arg parser and `MakeFirstValueReducer` factory, registered `FIRST_VALUE` in `GroupBy::reducerTable` with nargs 1–4.
- Added FIRST_VALUE to ft.aggregate.json and documented all three modes in `ft.aggregate.md`

## Testing
- `ft_aggregate_exec_test.cc`: covers simple mode, sorted ASC/DESC, case-insensitive keywords, and parse-time errors for invalid nargs, unknown BY token, and bad direction
- `ft_aggregate_exec_test.cc`: edge cases cover empty group, all-nil comparisons, nil return value preservation, tie-breaking, and mixed nil/non-nil inputs
- `test_non_vector.py`: end-to-end tests for simple mode, sorted ASC/DESC, multi-reducer with COUNT, and numeric self-comparison
- `generate.py`: added compatibility test -> `test_first_value_by_clause` for BY+ASC/DESC, default order, string fields, and cross-field return, simple mode excluded as non-deterministic.

## Example
```
FT.AGGREGATE products "@price:[1 1000]"
  LOAD 3 @price @rating @category
  GROUPBY 1 @category
  REDUCE FIRST_VALUE 4 @price BY @rating DESC AS price_with_max_rating
 ```
Returns, for each category, the price of the item with the highest rating. With ASC, returns the price of the item with the lowest rating.

### CODE REVIEW

File: docs/commands/ft.aggregate.md [hunk 1]
```diff
@@ -92,3 +92,6 @@ The following reducer functions are available. The reducer functions that take a
 | MAX 1 <expression>            | The largest numerical values of the expression.                                                                                    |
 | AVG 1 <expression>            | The numerical average of the values of the expression.                                                                             |
 | STDDEV 1 <expression>         | The standard deviation the values of the expression.                                                                               |
+| FIRST_VALUE 1 <expression>    | The first value of the expression encountered in the group. Order depends on record retrieval order and is non-deterministic. Use only when order does not matter. |
```
Comment (allenss-amazon) [inline docs/commands/ft.aggregate.md]: I'd drop the comments about non-deterministic. If a previous stage did a sort then it would be deterministic.

File: docs/commands/ft.aggregate.md [hunk 2]
```diff
@@ -92,3 +92,6 @@ The following reducer functions are available. The reducer functions that take a
...
+| FIRST_VALUE 3 <expression> BY <expression> | The value of the first expression from the record with the smallest comparison expression (ascending). Ties broken by first-encountered order. |
+| FIRST_VALUE 4 <expression> BY <expression> ASC\|DESC | The value of the first expression from the record with the minimum (ASC) or maximum (DESC) comparison expression. Ties broken by first-encountered order. Invalid keyword arguments (e.g., wrong BY token or unrecognised direction) produce a nil result. |
```
Comment (allenss-amazon) [inline docs/commands/ft.aggregate.md]: > Invalid keyword arguments (e.g., wrong BY token or unrecognised direction) produce a nil result

Why doesn't this generate a compile-time parse error?

File: integration/compatibility/generate.py [hunk 3]
```diff
@@ -463,8 +463,231 @@ def test_search_sortby(self, key_type, dialect):
...
+            f"reduce first_value 4 @t1 BY @n1 ASC as first_t1_by_n1"
+        )
+        
+        # Test comparing property to itself
+        self.check(dialect, 
+            f"ft.aggregate {key_type}_idx1 * "
+            f"load 3 @__key @n1 @n2 "
+            f"groupby 1 @n2 "
+            f"reduce first_value 4 @n1 BY @n1 ASC as first_n1_by_n1"
+        )
```
Comment (allenss-amazon) [inline integration/compatibility/generate.py]: Maybe some nested loops would be a better way to write this code.

File: integration/compatibility_test.py [hunk 4]
```diff
@@ -490,6 +490,20 @@ def do_answer_cluster(cluster_client, expected, data_set, test_case):
...
+    
+    answer_time = subprocess.run(["git", "log", "-1", "--format=%ct", answer_path], 
+                                  capture_output=True, text=True, cwd=root_dir).stdout.strip().split('\n')[-1]
+    generator_time = subprocess.run(["git", "log", "-1", "--format=%ct", generator_path], 
+                                     capture_output=True, text=True, cwd=root_dir).stdout.strip().split('\n')[-1]
+    
+    if not answer_time or (generator_time and int(generator_time) > int(answer_time)):
+        print(f"Regenerating {answer_file}...")
+        os.system(f"cd {root_dir}/integration/compatibility && pytest {generator_file}")
+
```
Comment (allenss-amazon) [inline integration/compatibility_test.py]: This code belongs in the cmake files, not here.

File: integration/test_non_vector.py [hunk 5]
```diff
@@ -459,6 +459,157 @@ def validate_aggregate_complex_queries(client: Valkey):
     assert result[1][1] == b'406'
     assert result[1][3] == b'4060'
 
+    # 17. FIRST_VALUE reducer - simple mode (no BY clause)
+    # Note: Simple mode is non-deterministic as it depends on retrieval order
```
Comment (allenss-amazon) [inline integration/test_non_vector.py:540]: What does this test cover that the compatibility tests didn't cover?

File: src/commands/ft_aggregate_exec.cc [hunk 6]
```diff
@@ -35,15 +35,15 @@ DEV_INTEGER_COUNTER(agg_stats, agg_sort_by_records);
 namespace valkey_search {
 namespace aggregate {
 
-expr::Value Attribute::GetValue(expr::Expression::EvalContext& ctx,
```
Comment (allenss-amazon) [inline src/commands/ft_aggregate_exec.cc:38]: let's get the code formatter aligned so that we don't see these huge meaningless changes.

If there's something about the formatter that's not being set in the build environment, let's fix that.

Comment (allenss-amazon) [reply src/commands/ft_aggregate_exec.cc:38]: I'll re-review when the diff covers the actual differences. It's too hard to tell now.

File: src/commands/ft_aggregate_parser.cc [hunk 7]
```diff
@@ -232,14 +232,21 @@ ConstructGroupByParser() {
...
-            VMSDK_ASSIGN_OR_RETURN(
-                auto expr,
-                expr::Expression::Compile(parameters, vmsdk::ToStringView(arg)),
-                _ << " in GROUPBY stage");
-            r.args_.emplace_back(std::move(expr));
+
+          if (r.info_->parse_args) {
+            VMSDK_ASSIGN_OR_RETURN(r.args_,
+                                   r.info_->parse_args(parameters, itr, cnt),
+                                   _ << " in GROUPBY stage");
```
Comment (allenss-amazon) [inline src/commands/ft_aggregate_parser.cc]: Maybe say "while parsing GROUPBY stage for function " << name

File: src/commands/ft_aggregate_parser.h [hunk 8]
```diff
@@ -224,14 +226,25 @@ class GroupBy : public Stage {
...
-    std::unique_ptr<ReducerInstance> (*make_instance)();
+    std::unique_ptr<ReducerInstance> (*make_instance)(
+        const std::vector<std::unique_ptr<expr::Expression>>& args);
+
+    // Optional custom argument parser. When set, called instead of the generic
+    // expression-compile loop. Receives the raw iterator positioned just after
+    // the nargs count token, and the declared nargs. Returns the compiled args
+    // to store in Reducer::args_, or an error surfaced at parse time.
+    using ArgParser =
+        absl::StatusOr<std::vector<std::unique_ptr<expr::Expression>>>(
```
Comment (allenss-amazon) [inline src/commands/ft_aggregate_parser.h]: why unique_ptr<expr> ? Why not just a vector of expr?

**Review by allenss-amazon — COMMENTED** [review]
love the rename of sortby into sortby_parameter.

File: src/commands/ft_aggregate_parser.h (hunk 8 above)
Comment (rileydes-improving) [reply src/commands/ft_aggregate_parser.h]: Expression is an abstract class, Compile() returns unique_ptr<Expression> and the type is only known at compile time. We can't store by value directly without slicing, I think unique_ptr is right here

File: integration/test_non_vector.py (hunk 5 above)
Comment (rileydes-improving) [reply integration/test_non_vector.py:540]: The compatibility testing skips simple mode because the compatibility tests require deterministic results against Redis. This test covers that gap, verifying that the reducer executes and returns a valid group member, ensuring that the simple mode has test coverage

File: integration/compatibility/generate.py [hunk 9]
```diff
@@ -463,8 +463,129 @@ def test_search_sortby(self, key_type, dialect):
...
                             self.check(dialect, f"ft.search {key_type}_idx1 * SORTBY {sort_key} {direction} {return_keys} {limit} {wsk}")
 
+
+    # test_first_value_simple_mode is intentionally omitted.
+    # FIRST_VALUE without a BY clause is non-deterministic: the order of
+    # records within a group depends on retrieval order, which differs between
+    # Redis and Valkey implementations. Compatibility testing requires
+    # deterministic results, so only BY-clause (sorted) mode is tested here.
+
+    def test_first_value_by_clause(self, key_type, dialect):
```
Comment (allenss-amazon) [inline integration/compatibility/generate.py:618]: Need to add tests for strings not just numbers.

**Comment by allenss-amazon** [comment]
Can we resolve the conflicts on this?

**Comment by rileydes-improving** [comment]
@allenss-amazon Rebased to current, resolved all conflicts, ready for review.

File: src/commands/ft_aggregate_exec.cc [hunk 10]
```diff
@@ -329,6 +329,48 @@ class Stddev : public GroupBy::ReducerInstance {
...
+    is_sorted_ = true;
+    is_desc_ = is_desc;
+  }
+
+  void ProcessRecord(const ArgVector& values) override {
+    if (!is_sorted_) {
+      // Simple mode: first record wins unconditionally.
+      if (result_value_.IsNil()) {
+        result_value_ = values[0];
+      }
```
[the remark this answers is not in the document]
Comment (allenss-amazon) [reply src/commands/ft_aggregate_exec.cc:352]: Let's add this test case to the compatibility test suite -- unless it's already there.

File: src/commands/ft_aggregate_exec.cc [hunk 11]
```diff
@@ -390,10 +434,104 @@ absl::StatusOr<std::unique_ptr<GroupBy::Reducer>> BasicReducerParser(
...
+                                              cnt, ") to reducer ", name));
+  }
+
+  // arg 0: the field whose value to return.
+  VMSDK_ASSIGN_OR_RETURN(auto field_tok, itr.PopNext(),
+                         _ << "Missing Reducer argument 0");
+  VMSDK_ASSIGN_OR_RETURN(
+      auto field_expr,
+      expr::Expression::Compile(parameters, vmsdk::ToStringView(field_tok)),
+      _ << " in GROUPBY stage");
```
Comment (allenss-amazon) [inline src/commands/ft_aggregate_exec.cc]: Diagnostic should be "In FIRST_VALUE reducer". not "in GROUPBY stage".

File: src/commands/ft_aggregate_exec.cc [hunk 12]
```diff
@@ -390,10 +434,104 @@ absl::StatusOr<std::unique_ptr<GroupBy::Reducer>> BasicReducerParser(
...
+  r->args_.push_back(std::move(field_expr));
+
+  if (cnt >= 3) {
+    // Expect "BY" keyword.
+    VMSDK_ASSIGN_OR_RETURN(auto by_tok, itr.PopNext(),
+                           _ << "Missing BY keyword");
+    auto by_upper = expr::FuncUpper(expr::Value(vmsdk::ToStringView(by_tok)));
```
Comment (allenss-amazon) [inline src/commands/ft_aggregate_exec.cc]: Should be using itr.PopIfNextIgnoreCase("BY") here.

File: src/commands/ft_aggregate_exec.cc [hunk 13]
```diff
@@ -390,10 +434,104 @@ absl::StatusOr<std::unique_ptr<GroupBy::Reducer>> BasicReducerParser(
...
+      VMSDK_ASSIGN_OR_RETURN(auto dir_tok, itr.PopNext(),
+                             _ << "Missing direction after sort field");
+      auto dir_upper =
+          expr::FuncUpper(expr::Value(vmsdk::ToStringView(dir_tok)));
+      auto dir_str = dir_upper.AsStringView();
+      if (dir_str != "ASC" && dir_str != "DESC") {
+        return absl::InvalidArgumentError(
+            absl::StrCat("FIRST_VALUE: expected ASC or DESC, got `",
+                         vmsdk::ToStringView(dir_tok), "`"));
+      }
```
Comment (allenss-amazon) [inline src/commands/ft_aggregate_exec.cc]: Same, use itr.PopIfNextIgnoreCase(...)

File: src/commands/ft_aggregate_exec.cc [hunk 14]
```diff
@@ -390,10 +434,104 @@ absl::StatusOr<std::unique_ptr<GroupBy::Reducer>> BasicReducerParser(
...
+          "FIRST_VALUE: expected BY, got `", vmsdk::ToStringView(by_tok), "`"));
+    }
+
+    // arg 1: the field to sort by.
+    VMSDK_ASSIGN_OR_RETURN(auto sort_tok, itr.PopNext(),
+                           _ << "Missing sort field after BY");
+    VMSDK_ASSIGN_OR_RETURN(
+        auto sort_expr,
+        expr::Expression::Compile(parameters, vmsdk::ToStringView(sort_tok)),
+        _ << " in GROUPBY stage");
```
Comment (allenss-amazon) [inline src/commands/ft_aggregate_exec.cc]: Diagnostic comment => in FIRST_VALUE....

**Review by allenss-amazon — CHANGES_REQUESTED** [review]
Some minor nits and then we can merge.

File: src/commands/ft_aggregate_exec.cc (hunk 14 above)
Comment (rileydes-improving) [reply src/commands/ft_aggregate_exec.cc]: Fixed in a2fa3956a1f07f67f679173e818fd5ae165252f1

File: src/commands/ft_aggregate_exec.cc (hunk 13 above)
Comment (rileydes-improving) [reply src/commands/ft_aggregate_exec.cc]: Fixed in a2fa3956a1f07f67f679173e818fd5ae165252f1

File: src/commands/ft_aggregate_exec.cc (hunk 12 above)
Comment (rileydes-improving) [reply src/commands/ft_aggregate_exec.cc]: Fixed in a2fa3956a1f07f67f679173e818fd5ae165252f1

File: src/commands/ft_aggregate_exec.cc (hunk 11 above)
Comment (rileydes-improving) [reply src/commands/ft_aggregate_exec.cc]: fixed in a2fa3956a1f07f67f679173e818fd5ae165252f1

File: integration/compatibility/generate.py (hunk 9 above)
Comment (rileydes-improving) [reply integration/compatibility/generate.py:618]: Added in a6fd0631297716939f09dbc7ad41160a85b74c89 and regenerated pickles.

**Review by allenss-amazon — APPROVED** [review]
Let's fix the merge conflicts and then will merge & close this.

**Comment by rileydes-improving** [comment]
> Let's fix the merge conflicts and then will merge & close this.

@allenss-amazon Fixed merge conflicts, ready to merge.

**Comment by Aksha1812** [comment]
**Reviewers for this PR**

- **First Pass Reviewer: @zackcam** — Please do your best to do a detailed review on the PR and get a response on your feedback. Once the first pass is done, notify the maintainer assigned to this PR to follow up on the final review and getting the PR merged. You can reach out to the people owning the relevant code paths for more help on the review.
- **Maintainer Reviewer: @BCathcart** — Once the first review is done, please follow up with a final review and help to merge the change in.
